docs/spec/store/README.md

```diff
@@ -0,0 +1,170 @@
+
+### `iavl.Store`
+
+The `iavl.Store` provides the core implementation for state storage and commitment
+by implementing the following interfaces:
+
+* `KVStore`
+* `CommitStore`
+* `CommitKVStore`
+* `Queryable`
+* `StoreWithInitialVersion`
+
+It allows for all CRUD operations to be performed along with allowing current
+and historical state queries, prefix iteration, and state commitment along with
+Merkle proof operations. The `iavl.Store` also provides the ability to remove historical state from the state commitment layer.
+
+An overview of the IAVL implementation can be found [here](https://github.com/cosmos/iavl/blob/master/docs/overview.md). It is important to note that the IAVL store
+provides both state commitment and logical storage operations, which comes with
+drawbacks as there are various performance impacts, some of which are very drastic,
+when it comes to the operations mentioned above.
+
+When dealing with state management in modules and clients, the Cosmos SDK provides
+various layers of abstractions or "store wrapping", where the `iavl.Store` is the
+bottom most layer. When requesting a store to perform reads or writes in a module,
+the typical abstraction layer in order is defined as follows:
+
+```text
+iavl.Store <- cachekv.Store <- gaskv.Store <- cachemulti.Store <- rootmulti.Store
+```
+
+### `cachekv.Store`
+
+The `cachekv.Store` store wraps an underlying `KVStore`, typically a `iavl.Store`
+and contains an in-memory cache for storing pending writes to underlying `KVStore`.
+`Set` and `Delete` calls are executed on the in-memory cache, whereas `Has` calls
+are proxied to the underlying `KVStore`. 
+
+One of the most important calls to a `cachekv.Store` is `Write()`, which ensures
+that key-value pairs are written to the underlying `KVStore` in a deterministic
+and ordered manner by sorting the keys first. The store keeps track of "dirty"
+keys and uses these to determine what keys to sort. In addition, it also keeps
+track of deleted keys and ensures these are also removed from the underlying
+`KVStore`.
+
+The `cachekv.Store` also provides the ability to perform iteration and reverse
+iteration. Iteration is performed through the `cacheMergeIterator` type and uses
+both the dirty cache and underlying `KVStore` to iterate over key-value pairs.
+
+Note, all calls to CRUD and iteration operations on a `cachekv.Store` are thread-safe.
+
+### `gaskv.Store`
+
+The `gaskv.Store` store provides a simple implementation of a `KVStore`.
+Specifically, it just wraps an existing `KVStore`, such as a cache-wrapped
+`iavl.Store`, and incurs configurable gas costs for CRUD operations via
+`ConsumeGas()` calls defined on the `GasMeter` which exists in a `sdk.Context`
+and then proxies the underlying CRUD call to the underlying store. Note, the
+`GasMeter` is reset on each block.
+
+### `cachemulti.Store` & `rootmulti.Store`
+
+The `rootmulti.Store` acts as an abstraction around a series of stores. Namely,
```

Comment:
lets open some issues to follow up,

what is a root multi store?

Reply:
What do you mean exactly? You think it's not described well enough here?

Reply:
how is it constructed? does this influence the app hash, if so how?

Reply:
This is already described. The doc states it computes a root Merkle application hash. I can add further notes on how it's constructed.